README
NAME
    AnyEvent::CurrentCost - AnyEvent module for reading from Current Cost
    energy meters

VERSION
    version 1.142000

SYNOPSIS
      # Create simple Current Cost reader with logging callback
      AnyEvent::CurrentCost->new(callback => sub { print $_[0]->summary },
                                 device => '/dev/ttyUSB0');

      # start event loop
      AnyEvent->condvar->recv;

DESCRIPTION
    AnyEvent module for reading from Current Cost energy meters.

    IMPORTANT: This is an early release and the API is still subject to
    change.

METHODS
  "new(%params)"
    Constructs a new "AnyEvent::CurrentCost" object. The supported
    parameters are:

    device
        The name of the device to connect to. The value should be a tty
        device name. The default is "/dev/ttyUSB0".

    callback
        The callback to execute when a message is received.

    history_callback
        A function, taking a sensor id, a time interval and a hash reference
        of data as arguments, to be called every time a new complete set of
        history data becomes available. The data hash reference has keys of
        the number of intervals ago and values of the reading at that time.

  "cleanup()"
    This method attempts to destroy any resources in the event of a
    disconnection or fatal error.

  "open()"
    This method opens the serial port and configures it.

  "anyevent_read_type()"
    This method is used to register an AnyEvent::Handle read type method to
    read Current Cost messages.

AUTHOR
    Mark Hindess <soft-cpan@temporalanomaly.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Mark Hindess.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension AnyEvent::CurrentCost.

1.142000  2014-07-19 14:25:01+01:00 Europe/London

1.130190  2013-01-19 17:38:04 Europe/London
    - Fix unstable test.
    - Use AnyEvent::SerialPort.

1.112970  2011-10-24 22:25:54 Europe/London
    - Fix prereqs.

1.110792  2011-03-20 13:07:21 Europe/London
    - More discard timeout fixes.

1.110791  2011-03-20 12:12:41 Europe/London
    - Avoid pointless discard timeout.

1.110790  2011-03-20 11:55:39 Europe/London
    - Initial release.


t/Helpers.pm
package t::Helpers;

=head1 NAME

t::Helpers - Perl extension for Helper functions for tests.

=head1 SYNOPSIS

  use Test::More tests => 2;
  use t::Helpers qw/:all/;
  is(test_error(sub { die 'argh' }),
     'argh',
     'died horribly');

  is(test_warn(sub { warn 'danger will robinson' }),
     'danger will robinson',
     'warned nicely');

=head1 DESCRIPTION

Common functions to make test scripts a bit easier to read.  There are
CPAN modules to do this sort of thing, but most people wont have them
installed and they are pretty trivial functions so to encourage
testing they are included here.

=cut

use strict;
use warnings;

use Exporter;

our @ISA = qw(Exporter);
our %EXPORT_TAGS = ( 'all' => [ qw(
                                    test_error
                                    test_warn
) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw();
our $VERSION = 0.1;

=head2 C<test_error($code_ref)>

This method runs the code with eval and returns the error.  It strips
off some common strings from the end of the message including any "at
<file> line <number>" strings and any "(@INC contains: .*)".

=cut

sub test_error {
  my $sub = shift;
  eval { $sub->() };
  my $error = $@;
  if ($error) {
    $error =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $error =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $error =~ s/ \(\@INC contains:.*?\)$//;
  }
  return $error;
}

=head2 C<test_warn($code_ref)>

This method runs the code with eval and returns the warning.  It strips
off any "at <file> line <number>" specific part(s) from the end.

=cut

sub test_warn {
  my $sub = shift;
  my $warn;
  local $SIG{__WARN__} = sub { $warn .= $_[0]; };
  eval { $sub->(); };
  die $@ if ($@);
  if ($warn) {
    $warn =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $warn =~ s/\s+at (\S+|\(eval \d+\)(\[[^]]+\])?) line \d+\.?\s*$//g;
    $warn =~ s/ \(\@INC contains:.*?\)$//;
  }
  return $warn;
}

1;


lib/AnyEvent/CurrentCost.pm
use strict;
use warnings;
package AnyEvent::CurrentCost;
$AnyEvent::CurrentCost::VERSION = '1.142000';
# ABSTRACT: AnyEvent module for reading from Current Cost energy meters


use constant DEBUG => $ENV{ANYEVENT_CURRENT_COST_DEBUG};
use base qw/Device::CurrentCost/;
use AnyEvent;
use AnyEvent::Handle;
use AnyEvent::SerialPort;
use Carp qw/croak carp/;
use Sub::Name;


sub new {
  my ($pkg, %p) = @_;
  croak $pkg.q{->new: 'callback' parameter is required} unless ($p{callback});
  my $self = $pkg->SUPER::new(%p);
  $self;
}

sub DESTROY { shift->cleanup }


sub cleanup {
  my $self = shift;
  print STDERR "cleanup\n" if DEBUG;
  delete $self->{handle};
  close $self->filehandle if (defined $self->filehandle);
}

sub _error {
  my ($self, $fatal, $message) = @_;
  $self->cleanup($message);
  $self->{on_error}->($fatal, $message) if ($self->{on_error});
}


sub open {
  my $self = shift;
  my $fh = $self->filehandle;
  my $handle =
    $fh
      ? AnyEvent::Handle->new(fh => $fh)
        : AnyEvent::SerialPort->new(serial_port =>
                                    [ $self->device,
                                      [ baudrate => $self->baud ] ]);
  print STDERR ref $self, "->open: created ", $handle, "\n" if DEBUG;
  $self->{handle} = $handle;
  $handle->on_error(subname 'on_error' => sub {
                      my ($handle, $fatal, $msg) = @_;
                      print STDERR $handle.": error $msg\n" if DEBUG;
                      $handle->destroy;
                      $self->_error($fatal, 'Error: '.$msg);
                    });
  $handle->on_rtimeout(subname 'on_rtimeout' => sub {
                         my $rbuf = \$handle->{rbuf};
                         carp $handle, ": Discarding '", $$rbuf, "'\n";
                         $$rbuf = '';
                         $handle->rtimeout(undef);
                       });
  $handle->on_read(subname 'on_read_cb' => sub {
                     my ($hdl) = @_;
                     $hdl->push_read(ref $self => $self,
                                     subname 'push_read_cb' => sub {
                                       $self->{callback}->(@_);
                                       1;
                                     });
                   });
}

sub _time_now {
  AnyEvent->now;
}


sub anyevent_read_type {
  my ($handle, $cb, $self) = @_;
  subname 'anyevent_read_type_reader' => sub {
    my $rbuf = \$handle->{rbuf};
    while (1) { # read all message from the buffer
      print STDERR "Before: ", (unpack 'H*', $$rbuf||''), "\n" if DEBUG;
      my $res = $self->read_one($rbuf);
      return unless ($res);
      print STDERR "After: ", (unpack 'H*', $$rbuf), "\n" if DEBUG;
      $handle->rtimeout($self->{discard_timeout}) if ($$rbuf && length $$rbuf);
      $res = $cb->($res);
    }
  }
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

AnyEvent::CurrentCost - AnyEvent module for reading from Current Cost energy meters

=head1 VERSION

version 1.142000

=head1 SYNOPSIS

  # Create simple Current Cost reader with logging callback
  AnyEvent::CurrentCost->new(callback => sub { print $_[0]->summary },
                             device => '/dev/ttyUSB0');

  # start event loop
  AnyEvent->condvar->recv;

=head1 DESCRIPTION

AnyEvent module for reading from Current Cost energy meters.

B<IMPORTANT:> This is an early release and the API is still subject to
change.

=head1 METHODS

=head2 C<new(%params)>

Constructs a new C<AnyEvent::CurrentCost> object.  The supported
parameters are:

=over

=item device

The name of the device to connect to.  The value should be a tty device
name.  The default is C</dev/ttyUSB0>.

=item callback

The callback to execute when a message is received.

=item history_callback

A function, taking a sensor id, a time interval and a hash reference
of data as arguments, to be called every time a new complete set of
history data becomes available.  The data hash reference has keys of
the number of intervals ago and values of the reading at that time.

=back

=head2 C<cleanup()>

This method attempts to destroy any resources in the event of a
disconnection or fatal error.

=head2 C<open()>

This method opens the serial port and configures it.

=head2 C<anyevent_read_type()>

This method is used to register an L<AnyEvent::Handle> read type
method to read Current Cost messages.

=head1 AUTHOR

Mark Hindess <soft-cpan@temporalanomaly.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Mark Hindess.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "AnyEvent module for reading from Current Cost energy meters",
  "AUTHOR" => "Mark Hindess <soft-cpan\@temporalanomaly.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "AnyEvent-CurrentCost",
  "EXE_FILES" => [
    "bin/anyevent-current-cost-reader"
  ],
  "LICENSE" => "perl",
  "NAME" => "AnyEvent::CurrentCost",
  "PREREQ_PM" => {
    "AnyEvent" => 0,
    "AnyEvent::Handle" => 0,
    "AnyEvent::SerialPort" => "1.130171",
    "Carp" => 0,
    "Device::CurrentCost" => "1.130190",
    "Device::CurrentCost::Constants" => 0,
    "Getopt::Long" => 0,
    "List::Util" => 0,
    "Pod::Usage" => 0,
    "Sub::Name" => 0,
    "base" => 0,
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "IO::Pipe" => 0,
    "Test::More" => 0,
    "Test::Requires" => 0,
    "Test::SharedFork" => 0,
    "Test::Warn" => 0
  },
  "VERSION" => "1.142000",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "AnyEvent" => 0,
  "AnyEvent::Handle" => 0,
  "AnyEvent::SerialPort" => "1.130171",
  "Carp" => 0,
  "Device::CurrentCost" => "1.130190",
  "Device::CurrentCost::Constants" => 0,
  "Getopt::Long" => 0,
  "IO::Pipe" => 0,
  "List::Util" => 0,
  "Pod::Usage" => 0,
  "Sub::Name" => 0,
  "Test::More" => 0,
  "Test::Requires" => 0,
  "Test::SharedFork" => 0,
  "Test::Warn" => 0,
  "base" => 0,
  "constant" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



